gensim/models/poincare.py

```diff
@@ -0,0 +1,803 @@
+#
+# Author: Jayant Jain <jayantjain1992@gmail.com>
+# Copyright (C) 2017 Radim Rehurek <me@radimrehurek.com>
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+
+"""
+Python implementation of Poincare Embeddings, an embedding to capture hierarchical information,
+described in [1]
+
+This module allows training a Poincare Embedding from a training file containing relations from
+a transitive closure.
+
+
+.. [1] https://arxiv.org/pdf/1705.08039.pdf
+
+"""
+
+
+import csv
+import itertools
+import logging
+import os
+import random
+import time
+
+from autograd import numpy as np, grad
+from collections import defaultdict, Counter
+from numpy import random as np_random
+from smart_open import smart_open
+
+from gensim import utils
+from gensim.models.keyedvectors import KeyedVectors, Vocab
+from gensim.models.word2vec import Word2Vec
+
+logger = logging.getLogger(__name__)
+
+
+class PoincareModel(utils.SaveLoad):
+    """
+    Class for training, using and evaluating Poincare Embeddings described in https://arxiv.org/pdf/1705.08039.pdf
+
+    The model can be stored/loaded via its `save()` and `load()` methods, or stored/loaded in the word2vec
+    format via `wv.save_word2vec_format()` and `KeyedVectors.load_word2vec_format()`.
+
+    Note that training cannot be resumed from a model loaded via `load_word2vec_format`, if you wish to train further,
+    use `save()` and `load()` methods instead.
+
+    """
+    def __init__(
+        self, train_data, size=50, alpha=0.1, negative=10, iter=50,
+        workers=1, epsilon=1e-5, burn_in=10, burn_in_alpha=0.01, seed=0):
+        """
+        Initialize and train a Poincare embedding model from an iterable of transitive closure relations.
+
+        Parameters
+        ----------
+        train_file : iterable
+            iterable of hypernym pairs, e.g. a list of tuples, or a PoincareData instance streaming from a file.
+        size : int, optional
+            Number of dimensions of the trained model.
+        alpha : float, optional
+            learning rate for training.
+        negative : int, optional
+            Number of negative samples to use.
+        iter : int, optional
+            Number of iterations (epochs) over the corpus.
+        workers : int, optional
+            Number of threads to use for training the model.
+        epsilon : float, optional
+            Constant used for clipping embeddings below a norm of one.
+        burn_in : int, optional
+            Number of epochs to use for burn-in initialization (0 means no burn-in).
+        burn_in_alpha : float, optional
+            learning rate for burn-in initialization, ignored if `burn_in` is 0.
+        seed : int, optional
+            seed for random to ensure reproducibility.
+
+        Returns
+        --------
+        PoincareModel instance
+
+        Examples
+        --------
+        Initialize a model from a list::
+
+        >>> model = PoincareModel(list_of_relations)
```

Comment:
It will be better if examples will be executable (i.e. with all needed imports + all variables defined)

Reply:
Added, also added working (although trivial) examples to the module docstring.